File: rfhe_framewk/src/ntt_test/relia_ntt_sim.py
import argparse, math, random
from typing import Dict, Tuple, List
import numpy as np

# ---------- 1. 参数配置 ----------

# 目标: 30-bit 素数, N=4096 (需要 q-1 能被 8192 整除以支持 4-step 的子变换)
TARGET_BIT_WIDTH = 30
N_DEFAULT = 4096

def is_prime(n):
    if n % 2 == 0: return False
    # 简单的素性测试 (Miller-Rabin for deterministic range or strict check)
    # 对于 30-bit 整数，简单试除或 python内置 pow 即可验证
    d = n - 1
    r = 0
    while d % 2 == 0:
        d //= 2
        r += 1
    for a in [2, 7, 61]: # bases for 32-bit int
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True

def find_ntt_prime(n_size, bit_width):
    # 寻找小于 2^bit_width 的最大 NTT 友好素数
    # q = k * 2N + 1 (为了安全通常找 2N，这里 N=4096, 找 k*8192+1)
    limit = 1 << bit_width
    step = 2 * n_size 
    start = (limit // step) * step + 1
    if start > limit: start -= step
    
    for q_cand in range(start, 1 << (bit_width-1), -step):
        if is_prime(q_cand):
            return q_cand
    raise ValueError("No suitable prime found")

def find_primitive_root(q):
    # 寻找原根
    if q == 2: return 1
    p1 = 2
    p2 = (q-1) // 2
    # 简单的假设 q = k*2^n + 1 的形式，尝试小整数
    for g in range(2, 100):
        # 原根的阶必须是 q-1
        # 对于 NTT 素数，主要是检查 g^((q-1)/2) != 1
        if pow(g, (q-1)//2, q) != 1:
            # 严谨的话要检查所有质因数，但对于 NTT 只要满足阶包含 N 即可
            # 这里简单返回一个生成元
            return g
    return 3

# 自动计算参数
q = find_ntt_prime(N_DEFAULT, TARGET_BIT_WIDTH)
g = find_primitive_root(q)

# 设置 24-bit 折叠模数 (2^24 + 1)
FOLD_MOD = (1 << 24) + 1 

BITS = q.bit_length()

# ---------- 2. 辅助函数 ----------
def mod_pow(a: int, e: int, m: int = q) -> int:
    return pow(a, e, m)

def root_of_unity(N: int) -> int:
    assert (q - 1) % N == 0, f"N={N} must divide q-1 ({q-1})"
    return mod_pow(g, (q - 1) // N, q)

def flip_bit_val(x: int, b: int) -> int:
    return (x ^ (1 << b)) % q

def flip_two_bits_val(x: int, b1: int, b2: int) -> int:
    if b1 == b2: return flip_bit_val(x, b1)
    return (x ^ (1 << b1) ^ (1 << b2)) % q

def inject_one(val: int, kind: str) -> int:
    """注入单一故障"""
    if kind == "SBF":
        b = random.randrange(BITS)
        return flip_bit_val(val, b)
    if kind == "DBF":
        b1 = random.randrange(BITS); b2 = random.randrange(BITS)
        return flip_two_bits_val(val, b1, b2)
    if kind == "MOF1":
        return random.randrange(q)
    raise ValueError("unknown fault kind")

# ---------- 3. 核心计算核心 (NTT & Twiddle) ----------

def ntt_inplace(vec: List[int], root: int,
                inj_plan: Dict[int, Tuple[str]] = None,
                op_base_idx: int = 0) -> int:
    """标准 NTT"""
    A = vec
    n = len(A)
    # 位反转
    j = 0
    for i in range(1, n):
        bit = n >> 1
        while j & bit:
            j ^= bit; bit >>= 1
        j ^= bit
        if i < j:
            A[i], A[j] = A[j], A[i]

    op_idx = op_base_idx
    length = 2
    while length <= n:
        wlen = mod_pow(root, n // length, q)
        for i in range(0, n, length):
            w = 1
            half = length // 2
            for j in range(i, i + half):
                u = A[j]
                # 蝶形乘法
                v_true = (A[j + half] * w) % q
                v = v_true
                
                if inj_plan and op_idx in inj_plan:
                    kind = inj_plan[op_idx][0]
                    v = inject_one(v, kind)
                op_idx += 1
                
                A[j]        = (u + v) % q
                A[j + half] = (u - v) % q
                w = (w * wlen) % q
        length <<= 1
    return op_idx

def twiddle_mul_inplace(M: np.ndarray, T: np.ndarray,
                        inj_plan: Dict[int, Tuple[str]] = None,
                        op_base_idx: int = 0) -> Tuple[int, bool]:
    """
    Twiddle 乘法 + Intra-element (24-bit Folding) 模拟
    """
    n2, n1 = M.shape
    op_idx = op_base_idx
    intra_detected_any = False 
    
    for r in range(n2):
        Tr = T[r]
        Mr = M[r]
        for c in range(n1):
            # 1. 真实计算
            v_true = (int(Mr[c]) * int(Tr[c])) % q
            v = v_true
            
            # 2. 注入故障
            if inj_plan and op_idx in inj_plan:
                kind = inj_plan[op_idx][0]
                v = inject_one(v, kind)
            op_idx += 1
            
            # 3. Intra-element 校验模拟 (Folding Check)
            if v != v_true:
                # 检查是否发生碰撞：模 (2^24 + 1)
                # 如果误差是 2^24+1 的倍数，这里会漏检 (return False)
                if (v % FOLD_MOD) != (v_true % FOLD_MOD):
                    intra_detected_any = True
                else:
                    # 只有 DBF(0, 24) 或类似组合会走到这里
                    # print(f"[DEBUG] Intra-check collision! v={v}, v_true={v_true}")
                    pass

            Mr[c] = v
            
    return op_idx, intra_detected_any

# ---------- 4. 保护校验逻辑 (ABFT) ----------

def batch_check_cols(A: np.ndarray, B: np.ndarray, n2_root: int) -> bool:
    """Batch-1 校验"""
    s_in  = np.sum(A, axis=1) % q
    s_out = np.sum(B, axis=1) % q
    w = [random.randrange(q) for _ in range(len(s_in))]
    w_hat = w.copy()
    ntt_inplace(w_hat, n2_root, None, 0)
    
    lhs = int(np.dot(w_hat, s_in) % q)
    rhs = int(np.dot(w,     s_out) % q)
    return lhs == rhs

def twiddle_check_inter(B_before: np.ndarray, B_after: np.ndarray, T: np.ndarray) -> bool:
    """Twiddle 阶段 Inter-element 校验"""
    n2, n1 = B_before.shape
    phi = np.array([random.randrange(q) for _ in range(n2)], dtype=np.int64)
    
    lhs = 0
    rhs = 0
    for c in range(n1):
        # 严格 ABFT 逻辑
        # lhs = <phi, Output>
        lhs = (lhs + int(np.dot(phi, B_after[:, c]) % q)) % q
        
        # rhs = <phi * T, Input>
        weighted_phi = (phi * T[:, c]) % q
        rhs = (rhs + int(np.dot(weighted_phi, B_before[:, c]) % q)) % q
        
    return lhs == rhs

def batch_check_rows(B: np.ndarray, C: np.ndarray, n1_root: int) -> bool:
    """Batch-2 校验"""
    r_in  = np.sum(B, axis=0) % q
    r_out = np.sum(C, axis=0) % q
    w = [random.randrange(q) for _ in range(len(r_in))]
    w_hat = w.copy()
    ntt_inplace(w_hat, n1_root, None, 0)
    
    lhs = int(np.dot(w_hat, r_in) % q)
    rhs = int(np.dot(w,     r_out) % q)
    return lhs == rhs

# ---------- 5. 四步法主流程 ----------

def four_step_ntt_protected(a: List[int], inj_plan: Dict[int, Tuple[str]] = None) -> Tuple[List[int], Dict]:
    N = len(a)
    n1 = int(round(math.isqrt(N))); n2 = n1
    assert n1 * n2 == N, "N must be a perfect square"

    wN   = root_of_unity(N)
    w_n1 = root_of_unity(n1)
    w_n2 = root_of_unity(n2)

    A = np.array([a[c*n2 + r] for c in range(n1) for r in range(n2)],
                 dtype=np.int64).reshape(n1, n2).T

    # Batch 1
    B = A.copy()
    op_idx = 0
    for c in range(n1):
        col = B[:, c].tolist()
        op_idx = ntt_inplace(col, w_n2, inj_plan, op_idx)
        B[:, c] = np.array(col, dtype=np.int64)
    ok_batch1 = batch_check_cols(A, B, w_n2)

    # Twiddle
    T = np.empty((n2, n1), dtype=np.int64)
    for r in range(n2):
        wr = mod_pow(wN, r, q)
        val = 1
        for c in range(n1):
            T[r, c] = val
            val = (val * wr) % q

    B_before = B.copy()
    op_idx, intra_detected = twiddle_mul_inplace(B, T, inj_plan, op_idx)
    inter_detected = not twiddle_check_inter(B_before, B, T)

    # Batch 2
    C = B.copy()
    for r in range(n2):
        row = C[r, :].tolist()
        op_idx = ntt_inplace(row, w_n1, inj_plan, op_idx)
        C[r, :] = np.array(row, dtype=np.int64)
    ok_batch2 = batch_check_rows(B, C, w_n1)

    out = C.T.flatten().astype(int).tolist()
    
    # 只要任意一个检测器触发就算成功
    is_detected = (not ok_batch1) or intra_detected or inter_detected or (not ok_batch2)
    
    info = dict(
        batch1_ok=int(ok_batch1), 
        intra_ok=int(not intra_detected), # 注意：这里 intra_detected=True 意味着发现了错误，所以 ok=False
        inter_ok=int(not inter_detected),
        batch2_ok=int(ok_batch2), 
        total_ops=op_idx,
        detected = int(is_detected)
    )
    return out, info

# ---------- 6. 实验 ----------

def make_single_injection_plan(total_ops: int, kind: str) -> Dict[int, Tuple[str]]:
    i = random.randrange(total_ops)
    return {i: (kind,)}

def run_trials(N: int, trials: int, kind: str, seed: int = 0) -> Tuple[float, float, Dict]:
    random.seed(seed)
    np.random.seed(seed)
    
    a0 = [0] * N
    _, info0 = four_step_ntt_protected(a0, inj_plan={})
    total_ops = info0["total_ops"]

    detected_count = 0
    # 统计是哪个机制抓住了错误
    stats = {"intra_catch": 0, "inter_catch": 0, "batch_catch": 0}

    for _ in range(trials):
        a = [random.randrange(q) for _ in range(N)]
        plan = make_single_injection_plan(total_ops, kind)
        _, info = four_step_ntt_protected(a, inj_plan=plan)
        
        if info["detected"]:
            detected_count += 1
            # 简单统计：优先归因为 Intra (最细粒度)，然后 Inter，然后 Batch
            if info["intra_ok"] == 0:
                stats["intra_catch"] += 1
            elif info["inter_ok"] == 0:
                stats["inter_catch"] += 1
            else:
                stats["batch_catch"] += 1
            
    det_rate = detected_count / trials
    miss_rate = 1.0 - det_rate
    return det_rate, miss_rate, stats

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--N", type=int, default=N_DEFAULT)
    ap.add_argument("--trials", type=int, default=10000)
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--kinds", type=str, default="SBF,DBF,MOF1")
    args = ap.parse_args()

    print(f"Configuration: N={args.N}")
    print(f"Found Prime q={q} (bit_len={q.bit_length()})")
    print(f"Intra-Fold-Mod={FOLD_MOD} (2^24 + 1)")
    
    kinds = [s.strip() for s in args.kinds.split(",") if s.strip()]
    
    for k in kinds:
        det, miss, stats = run_trials(args.N, args.trials, k, args.seed)
        print(f"Fault Type [{k}]: Det={det:.6f} Miss={miss:.6f}")
        print(f"    Breakdown: Intra={stats['intra_catch']}, Inter={stats['inter_catch']}, Batch={stats['batch_catch']}")

if __name__ == "__main__":
    main()

File: rfhe_framewk/src/ntt_test/relia_ntt_sim.cpp
// mc_ntt_multi_stages_qbits.cpp
// ReliaFHE Simulation - Multi-Fault Protocols with Stage-wise Breakdown & Configurable Q
// Features: SBF/DBF/MLF Combinations, Stage-specific Injection, Detailed Detection Stats, Q-Bits Config

#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <cstdint>
#include <algorithm>
#include <map>
#include <iomanip>
#include <omp.h>
#include <cassert>
#include <atomic>
#include <cstring>
#include <string>
#include <set>

using u64 = uint64_t;
using u128 = __uint128_t;

// ---------- Configuration (Globals) ----------
int N_GLOBAL = 4096;
int FOLD_BITS_GLOBAL = 4;
int TARGET_BIT_WIDTH_GLOBAL = 30; // Default 30-bit

// Global computed params
u64 Q = 0;
u64 G = 0;
int BITS = 0;
u64 FOLD_MASK = 0;

// ---------- Helpers ----------
u64 mod_pow(u64 base, u64 exp, u64 mod) {
    u64 res = 1;
    base %= mod;
    while (exp > 0) {
        if (exp % 2 == 1) res = (u128)res * base % mod;
        base = (u128)base * base % mod;
        exp /= 2;
    }
    return res;
}

bool is_prime(u64 n) {
    if (n < 2) return false;
    if (n == 2 || n == 3) return true;
    if (n % 2 == 0) return false;
    u64 d = n - 1;
    int r = 0;
    while (d % 2 == 0) { d /= 2; r++; }
    for (u64 a : {2, 7, 61}) {
        if (a >= n) break;
        u64 x = mod_pow(a, d, n);
        if (x == 1 || x == n - 1) continue;
        bool passed = false;
        for (int i = 0; i < r - 1; ++i) {
            x = (u128)x * x % n;
            if (x == n - 1) { passed = true; break; }
        }
        if (!passed) return false;
    }
    return true;
}

u64 find_ntt_prime(int n_size, int bit_width) {
    u64 limit = 1ULL << bit_width;
    u64 step = 2ULL * n_size; 
    u64 start = (limit / step) * step + 1;
    if (start > limit) start -= step;
    
    // Search downwards
    for (u64 q = start; q > (1ULL << (bit_width - 1)); q -= step) {
        if (is_prime(q)) return q;
    }
    
    // Fallback search up (for small bit widths)
    if (bit_width < 20) {
         start = (1ULL << (bit_width - 1)) / step * step + 1;
         for (u64 q = start; q < limit; q += step) {
             if (q > (1ULL << (bit_width - 1)) && is_prime(q)) return q;
         }
    }
    throw std::runtime_error("No suitable prime found");
}

u64 find_primitive_root(u64 q) {
    if (q == 2) return 1;
    for (u64 g = 2; g < 100; ++g) {
        if (mod_pow(g, (q - 1) / 2, q) != 1) return g;
    }
    return 3;
}

u64 root_of_unity(int N) {
    return mod_pow(G, (Q - 1) / N, Q);
}

// ---------- RNG & Injection ----------
struct ThreadRNG {
    std::mt19937_64 rng;
    ThreadRNG(u64 seed) : rng(seed) {}
    u64 rand_q() { return std::uniform_int_distribution<u64>(0, Q - 1)(rng); }
    u64 rand_full() { return rng(); }
    int rand_bits() { return std::uniform_int_distribution<int>(0, BITS - 1)(rng); }
};

enum FaultType {
    NONE = 0,
    SBF = 1,
    DBF = 2,
    MLF = 3
};

u64 flip_bit_val(u64 x, int b) { return (x ^ (1ULL << b)) % Q; }

u64 inject_one(u64 val, int kind, ThreadRNG& rng) {
    if (kind == SBF) { 
        return flip_bit_val(val, rng.rand_bits()); 
    }
    if (kind == DBF) { 
        int b1 = rng.rand_bits();
        int b2 = rng.rand_bits();
        while (b1 == b2) { b2 = rng.rand_bits(); } 
        return (val ^ (1ULL << b1) ^ (1ULL << b2)) % Q;
    }
    if (kind == MLF) { 
        return rng.rand_full() % Q; 
    }
    return val;
}

// ---------- Hardware Folding Logic ----------
inline u64 calc_fold_checksum(u64 val, int w) {
    u64 sum = 0;
    while (val > 0) {
        sum += (val & FOLD_MASK);
        val >>= w;
    }
    return sum & FOLD_MASK;
}

// ---------- Structures ----------
using InjectionPlan = std::vector<std::pair<long long, int>>;

struct OpRanges {
    long long s1_start; long long s1_end;
    long long s2_start; long long s2_end;
    long long s3_start; long long s3_end;
    long long total;
};

struct TrialStats {
    bool ok_b1;
    bool ok_intra;
    bool ok_inter;
    bool ok_b2;
    bool detected;
    OpRanges ranges;
};

struct TwiddleResult {
    long long ops;
    bool intra_detected;
};

// ---------- NTT Logic ----------
int ntt_inplace(std::vector<u64>& A, u64 root, 
                long long& op_idx, const InjectionPlan& plan, int& plan_ptr, ThreadRNG& rng) {
    int n = A.size();
    int j = 0;
    for (int i = 1; i < n; ++i) {
        int bit = n >> 1;
        while (j & bit) { j ^= bit; bit >>= 1; }
        j ^= bit;
        if (i < j) std::swap(A[i], A[j]);
    }

    int length = 2;
    while (length <= n) {
        u64 wlen = mod_pow(root, n / length, Q);
        for (int i = 0; i < n; i += length) {
            u64 w = 1;
            int half = length / 2;
            for (int k = i; k < i + half; ++k) {
                u64 u = A[k];
                u64 v_true = (u128)A[k + half] * w % Q;
                u64 v = v_true;

                // Check injection plan
                while (plan_ptr < plan.size() && plan[plan_ptr].first == op_idx) {
                    v = inject_one(v, plan[plan_ptr].second, rng);
                    plan_ptr++;
                }
                op_idx++;

                A[k] = (u + v) % Q;
                A[k + half] = (u + Q - v) % Q;
                w = (u128)w * wlen % Q;
            }
        }
        length <<= 1;
    }
    return 0; 
}

TwiddleResult twiddle_mul_inplace(std::vector<u64>& M, const std::vector<u64>& T, 
                                  long long start_op, const InjectionPlan& plan, int& plan_ptr, ThreadRNG& rng) {
    long long op_idx = start_op;
    bool intra_detected = false;
    
    for (size_t i = 0; i < M.size(); ++i) {
        u64 v_true = (u128)M[i] * T[i] % Q;
        u64 v = v_true;
        
        while (plan_ptr < plan.size() && plan[plan_ptr].first == op_idx) {
            v = inject_one(v, plan[plan_ptr].second, rng);
            plan_ptr++;
        }
        op_idx++;
        
        // Intra Check (Hardware Fold)
        if (v != v_true) {
            u64 chk_true = calc_fold_checksum(v_true, FOLD_BITS_GLOBAL);
            u64 chk_fault = calc_fold_checksum(v, FOLD_BITS_GLOBAL);
            if (chk_true != chk_fault) intra_detected = true;
        }
        M[i] = v;
    }
    return {op_idx, intra_detected};
}

// ---------- Protection Checks ----------
bool batch_check(const std::vector<u64>& In, const std::vector<u64>& Out, 
                 int n_sub, int len_sub, u64 root, ThreadRNG& rng, bool check_cols) {
    std::vector<u64> s_in(len_sub, 0);
    std::vector<u64> s_out(len_sub, 0);

    if (check_cols) { 
        int n1 = n_sub; int n2 = len_sub;
        for(int r=0; r<n2; ++r) {
            u128 sum_i=0, sum_o=0;
            for(int c=0; c<n1; ++c) {
                sum_i += In[c*n2+r]; sum_o += Out[c*n2+r];
            }
            s_in[r] = sum_i % Q; s_out[r] = sum_o % Q;
        }
    } else { 
        int n2 = n_sub; int n1 = len_sub;
        for(int c=0; c<n1; ++c) {
            u128 sum_i=0, sum_o=0;
            for(int r=0; r<n2; ++r) {
                sum_i += In[r*n1+c]; sum_o += Out[r*n1+c];
            }
            s_in[c] = sum_i % Q; s_out[c] = sum_o % Q;
        }
    }

    std::vector<u64> w(len_sub);
    for(int i=0; i<len_sub; ++i) w[i] = rng.rand_q();
    
    std::vector<u64> w_hat = w;
    // Dummy calls for w_hat transform
    InjectionPlan empty; int ptr=0; long long dummy_op=-1;
    ntt_inplace(w_hat, root, dummy_op, empty, ptr, rng);
    
    u128 lhs = 0, rhs = 0;
    for(int i=0; i<len_sub; ++i) {
        lhs = (lhs + (u128)w_hat[i] * s_in[i]) % Q;
        rhs = (rhs + (u128)w[i] * s_out[i]) % Q;
    }
    return lhs == rhs;
}

bool check_inter(const std::vector<u64>& B_pre, const std::vector<u64>& B_post, 
                 const std::vector<u64>& T, int n1, int n2, ThreadRNG& rng) {
    std::vector<u64> phi(n2);
    for(int i=0; i<n2; ++i) phi[i] = rng.rand_q();
    u128 lhs=0, rhs=0;
    for(int c=0; c<n1; ++c) {
        u128 lc=0, rc=0;
        for(int r=0; r<n2; ++r) {
            int idx = c*n2+r;
            lc = (lc + (u128)phi[r]*B_post[idx]) % Q;
            u64 w_phi = (u128)phi[r]*T[idx] % Q;
            rc = (rc + (u128)w_phi*B_pre[idx]) % Q;
        }
        lhs = (lhs+lc)%Q; rhs = (rhs+rc)%Q;
    }
    return lhs==rhs;
}

// ---------- 4-Step Flow ----------
TrialStats four_step_ntt(int N, const InjectionPlan& plan, ThreadRNG& rng) {
    int n1 = (int)round(sqrt(N));
    int n2 = N / n1;
    OpRanges r;
    int plan_ptr = 0;
    
    std::vector<u64> A(N);
    for(int i=0; i<N; ++i) A[i] = rng.rand_q();

    // Batch 1
    std::vector<u64> B = A;
    long long op_idx = 0;
    u64 w_n2 = root_of_unity(n2);
    
    r.s1_start = op_idx;
    for (int c = 0; c < n1; ++c) {
        std::vector<u64> col(B.begin() + c*n2, B.begin() + (c+1)*n2);
        ntt_inplace(col, w_n2, op_idx, plan, plan_ptr, rng);
        std::copy(col.begin(), col.end(), B.begin() + c*n2);
    }
    r.s1_end = op_idx;
    bool ok_b1 = batch_check(A, B, n1, n2, w_n2, rng, true);

    // Twiddle
    u64 wN = root_of_unity(N);
    std::vector<u64> T(N);
    for (int r = 0; r < n2; ++r) {
        u64 wr = mod_pow(wN, r, Q);
        u64 val = 1;
        for (int c = 0; c < n1; ++c) {
            T[c * n2 + r] = val;
            val = (u128)val * wr % Q;
        }
    }
    
    r.s2_start = op_idx;
    std::vector<u64> B_pre = B;
    TwiddleResult tw_res = twiddle_mul_inplace(B, T, op_idx, plan, plan_ptr, rng);
    op_idx = tw_res.ops;
    r.s2_end = op_idx;
    
    bool ok_inter = check_inter(B_pre, B, T, n1, n2, rng);

    // Batch 2
    std::vector<u64> C_transposed(N);
    for (int c = 0; c < n1; ++c) {
        for (int r = 0; r < n2; ++r) {
            C_transposed[r * n1 + c] = B[c * n2 + r];
        }
    }
    std::vector<u64> C_in = C_transposed;
    u64 w_n1 = root_of_unity(n1);
    
    r.s3_start = op_idx;
    for (int r = 0; r < n2; ++r) {
        std::vector<u64> row(C_transposed.begin() + r*n1, C_transposed.begin() + (r+1)*n1);
        ntt_inplace(row, w_n1, op_idx, plan, plan_ptr, rng);
        std::copy(row.begin(), row.end(), C_transposed.begin() + r*n1);
    }
    r.s3_end = op_idx;
    r.total = op_idx;
    
    bool ok_b2 = batch_check(C_in, C_transposed, n2, n1, w_n1, rng, false);

    bool detected = (!ok_b1) || tw_res.intra_detected || (!ok_inter) || (!ok_b2);
    return {ok_b1, !tw_res.intra_detected, ok_inter, ok_b2, detected, r};
}

// ---------- Main ----------
int main(int argc, char** argv) {
    int trials = 10000;
    
    for(int i=1; i<argc; i++) {
        std::string s = argv[i];
        if (s == "--trials" && i+1 < argc) trials = std::stoi(argv[++i]);
        else if (s == "--N" && i+1 < argc) N_GLOBAL = std::stoi(argv[++i]);
        else if (s == "--fold-bits" && i+1 < argc) FOLD_BITS_GLOBAL = std::stoi(argv[++i]);
        else if (s == "--q-bits" && i+1 < argc) TARGET_BIT_WIDTH_GLOBAL = std::stoi(argv[++i]);
    }

    FOLD_MASK = (1ULL << FOLD_BITS_GLOBAL) - 1;
    
    try {
        Q = find_ntt_prime(N_GLOBAL, TARGET_BIT_WIDTH_GLOBAL);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    
    G = find_primitive_root(Q);
    u64 tmp = Q; BITS = 0; while(tmp>0) { tmp>>=1; BITS++; }
    
    std::cout << "Config: N=" << N_GLOBAL << ", Fold=" << FOLD_BITS_GLOBAL 
              << " bits, Q=" << Q << " (" << BITS << "-bit)\n";
    std::cout << "Running Multi-Fault Protocol (Stage-wise)...\n";
    std::cout << std::string(95, '=') << std::endl;

    // Scenarios from Figure 10
    struct Scenario { 
        std::string name; 
        std::vector<int> faults; 
    };
    // std::vector<Scenario> scenarios = {
    //     {"SBF+SBF", {SBF, SBF}},
    //     {"SBF+DBF", {SBF, DBF}},
    // };
    std::vector<Scenario> scenarios = {
        {"SBF", {SBF}}, {"DBF", {DBF}}, {"MLF", {MLF}},
        {"SBF+SBF", {SBF, SBF}}, {"SBF+DBF", {SBF, DBF}},
        {"MLF+SBF", {MLF, SBF}}, {"MLF+DBF", {MLF, DBF}}, {"MLF+MLF", {MLF, MLF}}
    };

    // Dry Run for Ranges
    ThreadRNG drng(42);
    InjectionPlan empty;
    TrialStats dry = four_step_ntt(N_GLOBAL, empty, drng);
    OpRanges r = dry.ranges;
    
    struct StageDef { std::string name; long long start; long long end; };
    std::vector<StageDef> stages = {
        {"Stage1(Batch1)", r.s1_start, r.s1_end},
        {"Stage2(Twiddle)", r.s2_start, r.s2_end},
        {"Stage3(Batch2)", r.s3_start, r.s3_end}
    };

    for (auto& sc : scenarios) {
        std::cout << "Scenario: [" << sc.name << "]" << std::endl;
        std::cout << std::string(95, '-') << std::endl;
        std::cout << std::setw(18) << "Target Stage" 
                  << std::setw(12) << "Det Rate" 
                  << std::setw(12) << "Miss Rate"
                  << " | " 
                  << std::setw(10) << "Batch1" 
                  << std::setw(10) << "Intra" 
                  << std::setw(10) << "Inter" 
                  << std::setw(10) << "Batch2" << std::endl;

        for (auto& stg : stages) {
            std::atomic<long long> det_count{0};
            std::atomic<long long> c_b1{0}, c_intra{0}, c_inter{0}, c_b2{0};
            long long range_len = stg.end - stg.start;

            #pragma omp parallel
            {
                ThreadRNG rng(2025 + omp_get_thread_num());
                #pragma omp for
                for (int t = 0; t < trials; ++t) {
                    InjectionPlan plan;
                    std::set<long long> used_ops;
                    
                    // Generate Multi-Faults within the CURRENT Stage
                    for (int type : sc.faults) {
                        long long target;
                        do {
                            target = stg.start + (rng.rand_full() % range_len);
                        } while (used_ops.count(target));
                        used_ops.insert(target);
                        plan.push_back({target, type});
                    }
                    std::sort(plan.begin(), plan.end());
                    
                    TrialStats stats = four_step_ntt(N_GLOBAL, plan, rng);
                    
                    if (stats.detected) {
                        det_count++;
                        if (!stats.ok_b1) c_b1++;
                        if (!stats.ok_intra) c_intra++;
                        if (!stats.ok_inter) c_inter++;
                        if (!stats.ok_b2) c_b2++;
                    }
                }
            }

            double det_rate = (double)det_count / trials;
            double miss_rate = 1.0 - det_rate;
            
            std::cout << std::setw(18) << stg.name 
                      << std::fixed << std::setprecision(6) 
                      << std::setw(12) << det_rate 
                      << std::setw(12) << miss_rate
                      << " | " 
                      << std::setw(10) << c_b1 
                      << std::setw(10) << c_intra 
                      << std::setw(10) << c_inter 
                      << std::setw(10) << c_b2 << std::endl;
        }
        std::cout << std::string(95, '=') << std::endl;
    }
    
    return 0;
}
